### DISCUSSION TITLE
Is it possible to use Vike with another templating engine like Nunjucks ?

### DISCUSSION THREAD

**Discussion by ishaanparlikar** [vikejs/vike#1496] [Discussion]
Hi,
I was just wondering if is it possible to use something linke Nunjucks as templating engine using vike. We have to render our own templates written in nunjucks, pass data from our CMS and pre-render site.
Also is it possible to generate multiple routes programmatically from a single template?

**Comment by brillout** [comment]
In principe, you should be able to do that by using `onRenderHtml()` and `onRenderClient()`. Free to elaborate your use case.

**Reply by ishaanparlikar** [reply]
We have a script which turn our project structure into SSG site. It has njk templates which when run and passed different data will generate different html pages according to the config and specified routes.
I was planning to do the same. However, I wasn't able to see any config that does that like define a config file and pre-render will generate all the routes based on a template. 
Also, Is it necessary to use JS UI Frameworks?

Something like astro has. https://docs.astro.build/en/guides/routing/#static-ssg-mode

You pass different data to component to generate different pages from single template.

**Reply by brillout** [reply]
AFAICT you should be abe to do that and I don't foresee any blockers. See the pre-rendering and [render modes](https://vike.dev/render-modes) docs.

Before reaching out for help, it's best you try first, see https://brillout.github.io/rules/#try.
